ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,246 @@
+from ddtrace.internal.utils.formats import asbool
+from ddtrace.internal.utils.version import parse_version
+from ddtrace.internal.wrapping import unwrap
+from ddtrace.internal.wrapping import wrap
+from ddtrace.pin import Pin
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+graphql_version_str = getattr(graphql, "__version__")
+graphql_version = parse_version(graphql_version_str)
+
+if graphql_version < (3, 0):
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+config._add("graphql", dict(_default_service="graphql"))
+# Patching resolvers can produce a large trace, this environment variable will allow
+# users to disable this instrumentation
+_patch_resolvers = asbool(os.getenv("DD_TRACE_GRAPHQL_PATCH_RESOLVERS", default=True))
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+    setattr(graphql, "_datadog_patch", True)
+    Pin().onto(graphql)
+
+    if graphql_version < (3, 0):
+        to_patch = [
+            ("graphql.graphql", "graphql.query", "execute_graphql"),
+            ("graphql.language.parser", "graphql.parse", "parse"),
+            ("graphql.validation.validation", "graphql.validate", "validate"),
+            ("graphql.execution.executor", "graphql.execute", "execute"),
+        ]
+    else:
+        to_patch = [
+            ("graphql.graphql", "graphql.query", "graphql_impl"),
+            ("graphql.language.parser", "graphql.parse", "parse"),
+            ("graphql.validation.validate", "graphql.validate", "validate"),
+            ("graphql.execution.execute", "graphql.execute", "execute"),
+        ]
+
+    for module_str, span_name, patch_method in to_patch:
+        module = sys.modules[module_str]
+        func = getattr(module, patch_method)
+        wrapper = _traced_operation(span_name)
+        # wrap(...) updates the behavior of the original function.
+        # All references to the function will be patched.
+        wrap(func, wrapper)
+
+
+def unpatch():
+    if not getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    if graphql_version < (3, 0):
+        patched = [
+            ("graphql.graphql", "execute_graphql"),
+            ("graphql.language.parser", "parse"),
+            ("graphql.validation.validation", "validate"),
+            ("graphql.execution.executor", "execute"),
+        ]
+    else:
+        patched = [
+            ("graphql.graphql", "graphql_impl"),
+            ("graphql.language.parser", "parse"),
+            ("graphql.validation.validate", "validate"),
+            ("graphql.execution.execute", "execute"),
+        ]
+
+    for module_str, patch_method in patched:
+        module = sys.modules[module_str]
+        func = getattr(module, patch_method)
+        unwrap(func)
+
+    setattr(graphql, "_datadog_patch", False)
+
+
+def _traced_operation(span_name):
```

Comment:
With the new API of #3874 we probably want to cache this. That's because `unwrap` will need the same wrapper function as argument to work.